util/versionutils.c

```diff
@@ -0,0 +1,333 @@
+{
+  char *ver1, *ver2;
+  char *part1, *part2;
+  int index1 = 0, index2 = 0;
+  int release_state1 = 0, release_state2 = 0;
+  int rs1, rs2;
+
+  ver1 = prepare_version_string (version1);
+  ver2 = prepare_version_string (version2);
+
+  if (strcmp (ver1, ver2) == 0)
+    {
+      g_free (ver1);
+      g_free (ver2);
+      return (0);
+    }
+
+  if ((release_state1 = get_release_state (ver1, index1)))
+    index1++;
+  if ((release_state2 = get_release_state (ver2, index2)))
+    index2++;
+
+  part1 = get_part (ver1, index1);
+  part2 = get_part (ver2, index2);
+  while (part1 && part2)
+    {
+      if (strcmp (part1, part2) == 0)
+        {
+          index1++;
+          index2++;
+          g_free (part1);
+          g_free (part2);
+          part1 = get_part (ver1, index1);
+          part2 = get_part (ver2, index2);
+          continue;
+        }
+      else
+        break;
+    }
+
+  if (part1 == NULL && part2 == NULL)
+    return (release_state2 - release_state1);
+
+  if (is_text (part1) || is_text (part2))
+    {
+      if (part1)
+        g_free (part1);
+      if (part2)
+        g_free (part2);
+      return (-5); // undefined
+    }
+
+  rs1 = get_release_state (ver1, index1);
+  rs2 = get_release_state (ver2, index2);
+
+  if ((rs1 && release_state1) || (rs2 && release_state2))
+    return (-5); // undefined
+
+  if (part1 == NULL)
+    {
+      g_free (part2);
+      if (rs2)
+        return (rs2 - release_state1);
+      else
+        return (-1);
+    }
+
+  if (part2 == NULL)
+    {
+      g_free (part1);
+      if (rs1)
+        return (release_state2 - rs1);
+      else
+        return (1);
+    }
+
+  int ret = -5;
+
+  if (rs1 && rs2)
+    ret = rs2 - rs1;
+
+  if (rs1)
+    ret = -1;
+
+  if (rs2)
+    ret = 1;
+
+  if (!rs1 && !rs2 && atoi (part1) < atoi (part2))
+    ret = -1;
+
+  if (!rs1 && !rs2 && atoi (part1) == atoi (part2))
+    ret = 0;
+
+  if (!rs1 && !rs2 && atoi (part1) > atoi (part2))
+    ret = 1;
+
+  g_free (part1);
+  g_free (part2);
+  g_free (ver1);
+  g_free (ver2);
+  return (ret);
+}
+
+/**
+ * @brief Prepare the version string for comparison.
+ *
+ * @param[in]  version  The version string to generate the prepared
+ *                      version string from.
+ *
+ * @return  Returns a prepared copy of the version string version.
+ */
+static char *
+prepare_version_string (const char *version)
+{
+  char prep_version[1024];
+  char *ver;
+  int index_v, index_pv;
+  gboolean is_digit;
+
+  if (!version)
+    return (NULL);
+
+  ver = g_strdup (version);
+
+  /* set all characters to lowercase */
+  char *c = ver;
+  for (; *c; c++)
+    *c = tolower (*c);
+
+  index_v = index_pv = 0;
+
+  is_digit = g_ascii_isdigit (ver[0]);
+
+  while (index_v < (int) strlen (ver))
+    {
+      if (ver[index_v] == '_' || ver[index_v] == '-' || ver[index_v] == '+'
+          || ver[index_v] == ':' || ver[index_v] == '.')
+        {
+          if (index_pv > 0 && prep_version[index_pv - 1] != '.')
```

Comment:
I've only looked briefly but this looks like it could access -1 in prep_version when index_pv is 0.

Also prep_version is of limited size, so I'm guessing in general there should be size checks before it is accessed.